### PR TITLE
Implements @RootFragment annotation support.

### PR DESCRIPTION
**Pull request by smaugho**
I've implemented the support for a `@RootFragment` which solves the issue #1636 . Please let me know about any question/suggestion regarding the solution. 

Thanks.

### CODE REVIEW

File: AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/holder/EBeanHolder.java [hunk 1]
```diff
@@ -40,15 +40,18 @@
 
 	public static final String GET_INSTANCE_METHOD_NAME = "getInstance" + generationSuffix();
 
+	private JFieldVar rootFragmentField;
```
[the remark this answers is not in the document]
Comment (smaugho) [reply AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/holder/EBeanHolder.java]: all copyrights solved on this PR as well

File: AndroidAnnotations/androidannotations-core/androidannotations-api/src/main/java/org/androidannotations/annotations/RootFragment.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
+/**
+ * Copyright (C) 2010-2016 eBusiness Information, Excilys Group
```
[the remark this answers is not in the document]
Comment (smaugho) [reply AndroidAnnotations/androidannotations-core/androidannotations-api/src/main/java/org/androidannotations/annotations/RootFragment.java]: all copyrights solved on this PR as well

File: AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/core/handler/RootFragmentHandler.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+	private final InjectHelper<EBeanHolder> injectHelper;
+
+	public RootFragmentHandler(AndroidAnnotationsEnvironment environment) {
+		super(RootFragment.class, environment);
+		injectHelper = new InjectHelper<>(validatorHelper, this);
+	}
+
+	@Override
+	public void validate(Element element, ElementValidation validation) {
```
Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/core/handler/RootFragmentHandler.java]: I was wondering we should allow only Fragment extending types, however i see that you deliberately want to inject custom interfaces as well. `@RootContext` only allows types extending `Context` to be injected. I think we should do similar here.

**Comment by WonderCsabo** [comment]
@smaugho please add compile time tests to verify validation and that correct code compiles, and runtime tests as well.

File: AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/core/handler/RootFragmentHandler.java (hunk 3 above)
Comment (smaugho) [reply AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/core/handler/RootFragmentHandler.java]: Well, we could definitely check if it extends Fragment (support one, or system one)... I guess the annotation `@RootFragment` would be more clear then... I definitely would like to inject interfaces, though I could live with not being able to do it on this annotation ;)

In fact, I added an annotation `@RootView`  to the Plugin I'm building, which basically will use the "fragment or context activity" past to the EBean, and that one will support using Interfaces then. 

From an architecture point of view, it makes sense then, it would be quite usefull for implementing MVP and MVVM (though with some improvements to avoid holding a real instance of the activity/fragment on the beans). 

Well, this `@RootView` could be implemented directly on AA, but I think it will make more sense on this Plugin handling architectural patterns, but let me know if you think is useful on AA itself.

Comment (WonderCsabo) [reply AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/core/handler/RootFragmentHandler.java]: Let's require `Fragment` here then.

File: AndroidAnnotations/androidannotations-core/androidannotations/src/test/java/org/androidannotations/rootfragment/RootFragmentTest.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	}
+
+	@Test
+	public void eBeanOnCustomFragmentCompiles() {
+		assertCompilationSuccessful(compileFiles(SomeBeanWithRootFragmentWithCustomFragment.class));
+	}
+
+	@Test
+	public void eBeanOnCustomSupportFragmentsCompiles() {
+		assertCompilationSuccessful(compileFiles(defPath("support/Fragment.java", "CustomSupportFragment.java", "SomeBeanWithRootFragmentWithCustomSupportFragment.java")));
```
Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/test/java/org/androidannotations/rootfragment/RootFragmentTest.java]: Why is this needed? I see similar tests, for example `ActivityIntentFragmentTest`, which is using `toPath()`.

File: AndroidAnnotations/androidannotations-core/androidannotations-testutils/src/main/java/org/androidannotations/testutils/ProcessorTestHelper.java [hunk 5]
```diff
@@ -293,6 +293,10 @@ public CompileResult compileFiles(Object... elements) {
 				addCollection(files, (Type) element);
 			} else if (element instanceof String) {
 				files.add(new File((String) element));
+			} else if (element instanceof String[]) {
```
Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations-core/androidannotations-testutils/src/main/java/org/androidannotations/testutils/ProcessorTestHelper.java]: Why is this needed?

File: AndroidAnnotations/androidannotations-core/androidannotations/src/test/java/org/androidannotations/rootfragment/RootFragmentTest.java (hunk 4 above)
Comment (WonderCsabo) [reply AndroidAnnotations/androidannotations-core/androidannotations/src/test/java/org/androidannotations/rootfragment/RootFragmentTest.java]:
```java
CompileResult result = compileFiles(toPath(ActivityIntentFragmentTest.class, "support/Fragment.java"), //
				toPath(ActivityIntentFragmentTest.class, "support/ActivityCompat.java"), ActivityInManifest.class);
```

Why did you add the `defPath()` method?

File: AndroidAnnotations/androidannotations-core/androidannotations-test/src/test/java/org/androidannotations/test/rootfragment/RootFragmentTest.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+import org.junit.runner.RunWith;
+import org.robolectric.Robolectric;
+import org.robolectric.RobolectricTestRunner;
+
+@RunWith(RobolectricTestRunner.class)
+public class RootFragmentTest {
+
+	@Test
+	public void injectFragmentWhichInjectedBean() {
+		CustomFragment fragment = new CustomFragment_();
```
Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations-core/androidannotations-test/src/test/java/org/androidannotations/test/rootfragment/RootFragmentTest.java]: I would create the `Fragment` properly using `Robolectric`:
```java
CustomFragment fragment = Robolectric.buildFragment(CustomFragment_.class).create().get();
```

This applies to other test methods as well.

File: AndroidAnnotations/androidannotations-core/androidannotations-test/src/main/java/org/androidannotations/test/rootfragment/ActivityWithBeansWithRootFragment.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed To in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.androidannotations.test.rootfragment;
```
Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations-core/androidannotations-test/src/main/java/org/androidannotations/test/rootfragment/ActivityWithBeansWithRootFragment.java]: Please move all this tests under `ebean` package. (`@RootContext` is also tested under that).

File: AndroidAnnotations/androidannotations-core/androidannotations/src/test/java/org/androidannotations/rootfragment/CustomFragment.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed To in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.androidannotations.rootfragment;
```
Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/test/java/org/androidannotations/rootfragment/CustomFragment.java]: Please move all the compile-time tests as well under `ebean` package.

File: AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/holder/EBeanHolder.java [hunk 9]
```diff
@@ -58,6 +62,16 @@ private void setConstructor() {
...
 		}
 		constructorBody.assign(getContextField(), constructorContextParam);
+
+		overloadedConstructor = generatedClass.constructor(PRIVATE);
+		JVar overloadedConstructorContextParam = overloadedConstructor.param(getClasses().CONTEXT, "context");
+		JVar overloadedConstructorRootFragmentParam = overloadedConstructor.param(getClasses().OBJECT, "rootFragment");
+		JBlock overloadedConstructorBody = overloadedConstructor.body();
+		if (superConstructor.getParameters().size() == 1) {
+			overloadedConstructorBody.invoke("super").arg(constructorContextParam);
+		}
```
Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/holder/EBeanHolder.java]: Travis fails because of this i think. You have to write:
```java
overloadedConstructorBody.add(JExpr.invokeSuper().arg(constructorContextParam));
```

File: AndroidAnnotations/androidannotations-core/androidannotations/src/test/java/org/androidannotations/rootfragment/RootFragmentTest.java (hunk 4 above)
Comment (smaugho) [reply AndroidAnnotations/androidannotations-core/androidannotations/src/test/java/org/androidannotations/rootfragment/RootFragmentTest.java]: @WonderCsabo , "defPath" accepts multiple paths (which is needed when having several "support-dependent files".. Also, it simplifies having to pass "the current class" per each call to "toPath", it is a little improvement, without it, when you have something like three files (like in that example above), basically you have to make a huge line with "toPath" and passing the test class to that method  (so 3 times). That doesn't seem that clean. 

"def" just assumes that the "package-holder class" is the same than the one of the test

File: AndroidAnnotations/androidannotations-core/androidannotations-testutils/src/main/java/org/androidannotations/testutils/ProcessorTestHelper.java (hunk 5 above)
Comment (smaugho) [reply AndroidAnnotations/androidannotations-core/androidannotations-testutils/src/main/java/org/androidannotations/testutils/ProcessorTestHelper.java]: @WonderCsabo , this comes with the function I explained above, "defPath", which basically returns an array of strings (multiple paths), so on this one, that is simply parsed to get each one of the paths and add them to the files.

File: AndroidAnnotations/androidannotations-core/androidannotations-test/src/test/java/org/androidannotations/test/rootfragment/RootFragmentTest.java (hunk 6 above)
Comment (smaugho) [reply AndroidAnnotations/androidannotations-core/androidannotations-test/src/test/java/org/androidannotations/test/rootfragment/RootFragmentTest.java]: Yeap, I tried to follow how I saw it in others.. mm.. but if the Fragment is not "added" in any FragmentManager, I don't think it would make that difference using Robolectric for it... or is there any?

File: AndroidAnnotations/androidannotations-core/androidannotations-test/src/main/java/org/androidannotations/test/rootfragment/ActivityWithBeansWithRootFragment.java (hunk 7 above)
Comment (smaugho) [reply AndroidAnnotations/androidannotations-core/androidannotations-test/src/main/java/org/androidannotations/test/rootfragment/ActivityWithBeansWithRootFragment.java]: That explains me why I didn't see any for @RootContext, moving them will required a little refactorization 😅 .

File: AndroidAnnotations/androidannotations-core/androidannotations-test/src/test/java/org/androidannotations/test/rootfragment/RootFragmentTest.java (hunk 6 above)
Comment (WonderCsabo) [reply AndroidAnnotations/androidannotations-core/androidannotations-test/src/test/java/org/androidannotations/test/rootfragment/RootFragmentTest.java]: Robolectric is adding it to a `FragmentManager`, using the method i described above.

File: AndroidAnnotations/androidannotations-core/androidannotations/src/test/java/org/androidannotations/rootfragment/RootFragmentTest.java (hunk 4 above)
Comment (WonderCsabo) [reply AndroidAnnotations/androidannotations-core/androidannotations/src/test/java/org/androidannotations/rootfragment/RootFragmentTest.java]: I see it now, thanks for improving it!

File: AndroidAnnotations/androidannotations-core/androidannotations-testutils/src/main/java/org/androidannotations/testutils/ProcessorTestHelper.java (hunk 5 above)
Comment (WonderCsabo) [reply AndroidAnnotations/androidannotations-core/androidannotations-testutils/src/main/java/org/androidannotations/testutils/ProcessorTestHelper.java]: I understand, thanks.

File: AndroidAnnotations/androidannotations-core/androidannotations-test/src/test/java/org/androidannotations/test/rootfragment/RootFragmentTest.java (hunk 6 above)
Comment (smaugho) [reply AndroidAnnotations/androidannotations-core/androidannotations-test/src/test/java/org/androidannotations/test/rootfragment/RootFragmentTest.java]: Ah got it, if so then let's do it like that, since that will bring it closer to the real runtime scenario.

**Comment by smaugho** [comment]
@WonderCsabo Travis failed this time but seems to be due to some missing copyrights in some other files

**Comment by smaugho** [comment]
I will add them..

**Comment by WonderCsabo** [comment]
Thanks, that was my fault. I forgot to rebase the branch to include the license check...

**Comment by WonderCsabo** [comment]
BTW, next time please create separate commits for unrelated changes such as this.

**Comment by WonderCsabo** [comment]
This depends on https://github.com/androidannotations/androidannotations/pull/2211 , right?

**Comment by smaugho** [comment]
Right now yes 😅

**Comment by dodgex** [comment]
https://github.com/androidannotations/androidannotations/pull/2211 has been merged.

**Comment by smaugho** [comment]
I will rebase this branch, to simplify its revision.

**Comment by WonderCsabo** [comment]
@smaugho please rebase this.

**Comment by smaugho** [comment]
Done

**Comment by dodgex** [comment]
Thanks! :)
